// weboptimizer/ejsLoader.compiled.js
#!/usr/bin/env node


// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports

var _babelCore = require('babel-core');

var _babelPresetBabili = require('babel-preset-babili');

var _babelPresetBabili2 = _interopRequireDefault(_babelPresetBabili);

var _babelPluginTransformWith = require('babel-plugin-transform-with');

var _babelPluginTransformWith2 = _interopRequireDefault(_babelPluginTransformWith);

var _clientnode = require('clientnode');

var _clientnode2 = _interopRequireDefault(_clientnode);

var _ejs = require('ejs');

var ejs = _interopRequireWildcard(_ejs);

var _fs = require('fs');

var fileSystem = _interopRequireWildcard(_fs);

var _htmlMinifier = require('html-minifier');

var _loaderUtils = require('loader-utils');

var loaderUtils = _interopRequireWildcard(_loaderUtils);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _configurator = require('./configurator.compiled');

var _configurator2 = _interopRequireDefault(_configurator);

var _helper = require('./helper.compiled');

var _helper2 = _interopRequireDefault(_helper);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// NOTE: Only needed for debugging this file.
try {
    require('source-map-support/register');
} catch (error) {}
// endregion

// endregion
// region types
module.exports = function (source) {
    if ('cachable' in this && this.cacheable) this.cacheable();
    const query = _clientnode2.default.convertSubstringInPlainObject(_clientnode2.default.extendObject(true, {
        compress: {
            html: {},
            javaScript: {}
        },
        context: './',
        extensions: {
            file: {
                external: ['.js'],
                internal: ['.js', '.css', '.svg', '.png', '.jpg', '.gif', '.ico', '.html', '.json', '.eot', '.ttf', '.woff']
            }, module: []
        },
        module: {
            aliases: {},
            replacements: {}
        },
        compileSteps: 2
    }, this.options || {}, 'query' in this ? loaderUtils.getOptions(this) || {} : {}), /#%%%#/g, '!');
    const compile = (template, options = query.compiler, compileSteps = 2) => (locals = {}) => {
        options = _clientnode2.default.extendObject(true, { filename: template }, options);
        const require = (request, nestedLocals = {}) => {
            const template = request.replace(/^(.+)\?[^?]+$/, '$1');
            const queryMatch = request.match(/^[^?]+\?(.+)$/);
            if (queryMatch) {
                const evaluationFunction = (request, template, source, compile, locals) => new Function('request', 'template', 'source', 'compile', 'locals', `return ${queryMatch[1]}`)(request, template, source, compile, locals);
                nestedLocals = _clientnode2.default.extendObject(true, nestedLocals, evaluationFunction(request, template, source, compile, locals));
            }
            let nestedOptions = _clientnode2.default.copyLimitedRecursively(options);
            delete nestedOptions.client;
            nestedOptions = _clientnode2.default.extendObject(true, { encoding: _configurator2.default.encoding }, nestedOptions, nestedLocals.options || {});
            if (nestedOptions.isString) return compile(template, nestedOptions)(nestedLocals);
            const templateFilePath = _helper2.default.determineModuleFilePath(template, query.module.aliases, query.module.replacements, query.extensions, query.context, _configurator2.default.path.source.asset.base, _configurator2.default.path.ignore, _configurator2.default.module.directoryNames, _configurator2.default.package.main.fileNames, _configurator2.default.package.main.propertyNames, _configurator2.default.package.aliasPropertyNames, _configurator2.default.encoding);
            if (templateFilePath) {
                if ('query' in this) this.addDependency(templateFilePath);
                /*
                    NOTE: If there aren't any locals options or variables and
                    file doesn't seem to be an ejs template we simply load
                    included file content.
                */
                if (queryMatch || templateFilePath.endsWith('.ejs')) return compile(templateFilePath, nestedOptions)(nestedLocals);
                // IgnoreTypeCheck
                return fileSystem.readFileSync(templateFilePath, nestedOptions);
            }
            throw new Error(`Given template file "${template}" couldn't be resolved.`);
        };
        const compressHTML = content => query.compress.html ? (0, _htmlMinifier.minify)(content, _clientnode2.default.extendObject(true, {
            caseSensitive: true,
            collapseInlineTagWhitespace: true,
            collapseWhitespace: true,
            conservativeCollapse: true,
            minifyCSS: true,
            minifyJS: true,
            processScripts: ['text/ng-template', 'text/x-handlebars-template'],
            removeAttributeQuotes: true,
            removeComments: true,
            removeRedundantAttributes: true,
            removeScriptTypeAttributes: true,
            removeStyleLinkTypeAttributes: true,
            sortAttributes: true,
            sortClassName: true,
            // NOTE: Avoids whitespace around placeholder in tags.
            trimCustomFragments: true,
            useShortDoctype: true
        }, query.compress.html)) : content;
        let remainingSteps = compileSteps;
        let result = template;
        let isString = options.isString;
        delete options.isString;
        while (remainingSteps > 0) {
            if (typeof result === 'string') {
                const filePath = isString && options.filename || result;
                if (filePath && _path2.default.extname(filePath) === '.js') result = eval('require')(filePath);else {
                    if (!isString) {
                        let encoding = _configurator2.default.encoding;
                        if ('encoding' in options) encoding = options.encoding;
                        result = fileSystem.readFileSync(result, { encoding });
                    }
                    if (remainingSteps === 1) result = compressHTML(result);
                    result = ejs.compile(result, options);
                }
            } else result = compressHTML(result(_clientnode2.default.extendObject(true, {
                configuration: _configurator2.default, Helper: _helper2.default, include: require, require, Tools: _clientnode2.default
            }, locals)));
            remainingSteps -= 1;
        }
        if (Boolean(compileSteps % 2)) return `'use strict';\n` + (0, _babelCore.transform)(`module.exports = ${result.toString()};`, {
            ast: false,
            babelrc: false,
            comments: !Boolean(query.compress.javaScript),
            compact: Boolean(query.compress.javaScript),
            filename: options.filename || 'unknown',
            minified: Boolean(query.compress.javaScript),
            plugins: [_babelPluginTransformWith2.default],
            presets: query.compress.javaScript ? [[_babelPresetBabili2.default, query.compress.javaScript]] : [],
            sourceMaps: false,
            sourceType: 'script'
        }).code;
        // IgnoreTypeCheck
        return result;
    };
    return compile(source, {
        client: Boolean(query.compileSteps % 2),
        compileDebug: this.debug || false,
        debug: this.debug || false,
        filename: 'query' in this ? loaderUtils.getRemainingRequest(this).replace(/^!/, '') : this.filename || null,
        isString: true
    }, query.compileSteps)(query.locals || {});
};
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion
