// weboptimizer/browserAPI.compiled.js
// #!/usr/bin/env node

// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports

Object.defineProperty(exports, "__esModule", {
    value: true
});

// endregion
// region variables
const onCreatedListener = [];
// endregion
// region declaration

let browserAPI;
// endregion
// region ensure presence of common browser environment
if (typeof TARGET_TECHNOLOGY === 'undefined' || TARGET_TECHNOLOGY === 'node') {
    // region mock browser environment
    const path = require('path');
    const { JSDOM, VirtualConsole } = require('jsdom');
    const virtualConsole = new VirtualConsole();
    for (const name of ['assert', 'dir', 'info', 'log', 'time', 'timeEnd', 'trace', 'warn']) virtualConsole.on(name, console[name].bind(console));
    virtualConsole.on('error', function (error) {
        if (!browserAPI.debug && ['XMLHttpRequest', 'resource loading'
        // IgnoreTypeCheck
        ].includes(error.type)) console.warn(`Loading resource failed: ${error.toString()}.`);else
            // IgnoreTypeCheck
            console.error(error.stack, error.detail);
    });
    const render = function render(template) {
        let window = new JSDOM(template, {
            resources: 'usable',
            runScripts: 'dangerously',
            url: 'http://localhost',
            virtualConsole
        }).window;
        browserAPI = {
            debug: false, domContentLoaded: false, DOM: JSDOM, window,
            windowLoaded: false };
        window.addEventListener('load', function () {
            // NOTE: Maybe we have miss the "DOMContentLoaded" event.
            browserAPI.domContentLoaded = true;
            browserAPI.windowLoaded = true;
        });
        window.document.addEventListener('DOMContentLoaded', function () {
            browserAPI.domContentLoaded = true;
        });
        for (const callback of onCreatedListener) callback(browserAPI, false);
        return window;
    };
    if (typeof NAME === 'undefined' || NAME === 'webOptimizer') {
        const filePath = path.join(__dirname, 'index.html.ejs');
        require('fs').readFile(filePath, { encoding: 'utf-8' }, function (error, content) {
            if (error) throw error;
            render(require('./ejsLoader.compiled').bind({ filename: filePath })(content));
        });
    } else
        // IgnoreTypeCheck
        render(require('webOptimizerDefaultTemplateFilePath'));
    // endregion
} else {
    browserAPI = {
        debug: false, domContentLoaded: false, DOM: null, window,
        windowLoaded: false };
    window.document.addEventListener('DOMContentLoaded', function () {
        browserAPI.domContentLoaded = true;
        for (const callback of onCreatedListener) callback(browserAPI, false);
    });
    window.addEventListener('load', function () {
        browserAPI.windowLoaded = true;
    });
}
// endregion

exports.default = function (callback, clear = true) {
    // region initialize global context
    const wrappedCallback = function wrappedCallback(...parameter) {
        if (clear && typeof global !== 'undefined' && global !== browserAPI.window) global.window = browserAPI.window;
        return callback(...parameter);
    };
    // endregion
    if (typeof TARGET_TECHNOLOGY === 'undefined' || TARGET_TECHNOLOGY === 'node') return browserAPI ? wrappedCallback(browserAPI, true) : onCreatedListener.push(wrappedCallback);
    return browserAPI.domContentLoaded ? wrappedCallback(browserAPI, true) : onCreatedListener.push(wrappedCallback);
};
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion
