// weboptimizer/webpackConfigurator.compiled.js
#!/usr/bin/env node


// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _babiliWebpackPlugin = require('babili-webpack-plugin');

var _babiliWebpackPlugin2 = _interopRequireDefault(_babiliWebpackPlugin);

var _clientnode = require('clientnode');

var _clientnode2 = _interopRequireDefault(_clientnode);

var _jsdom = require('jsdom');

var _fs = require('fs');

var fileSystem = _interopRequireWildcard(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _postcssCssnext = require('postcss-cssnext');

var _postcssCssnext2 = _interopRequireDefault(_postcssCssnext);

var _postcssFontpath = require('postcss-fontpath');

var _postcssFontpath2 = _interopRequireDefault(_postcssFontpath);

var _postcssImport = require('postcss-import');

var _postcssImport2 = _interopRequireDefault(_postcssImport);

var _postcssSprites = require('postcss-sprites');

var _postcssSprites2 = _interopRequireDefault(_postcssSprites);

var _postcssUrl = require('postcss-url');

var _postcssUrl2 = _interopRequireDefault(_postcssUrl);

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _webpackSources = require('webpack-sources');

var _ejsLoader = require('./ejsLoader.compiled');

var _ejsLoader2 = _interopRequireDefault(_ejsLoader);

var _configurator = require('./configurator.compiled');

var _configurator2 = _interopRequireDefault(_configurator);

var _helper = require('./helper.compiled');

var _helper2 = _interopRequireDefault(_helper);

var _htmlLoader = require('html-loader');

var _htmlLoader2 = _interopRequireDefault(_htmlLoader);

var _loaderUtils = require('loader-utils');

var _loaderUtils2 = _interopRequireDefault(_loaderUtils);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// NOTE: Only needed for debugging this file.

/* eslint-enable no-unused-vars */

/* eslint-disable no-unused-vars */
try {
    require('source-map-support/register');
} catch (error) {}

const plugins = require('webpack-load-plugins')();


plugins.HTML = plugins.html;
plugins.ExtractText = plugins.extractText;
plugins.AddAssetHTMLPlugin = require('add-asset-html-webpack-plugin');
plugins.OpenBrowser = plugins.openBrowser;
plugins.Favicon = require('favicons-webpack-plugin');
plugins.Imagemin = require('imagemin-webpack-plugin').default;
plugins.Offline = require('offline-plugin');

// / region monkey patches
// Monkey-Patch html loader to retrieve html loader options since the
// "webpack-html-plugin" doesn't preserve the original loader interface.

require.cache[require.resolve('html-loader')].exports = function (...parameter) {
    _clientnode2.default.extendObject(true, this.options, module, this.options);
    return _htmlLoader2.default.call(this, ...parameter);
};
// Monkey-Patch loader-utils to define which url is a local request.

const loaderUtilsIsUrlRequestBackup = _loaderUtils2.default.isUrlRequest;
require.cache[require.resolve('loader-utils')].exports.isUrlRequest = (url, ...additionalParameter) => {
    if (url.match(/^[a-z]+:.+/)) return false;
    return loaderUtilsIsUrlRequestBackup.apply(_loaderUtils2.default, [url].concat(additionalParameter));
};
// / endregion
// endregion
// region initialisation
// / region determine library name
let libraryName;
if ('libraryName' in _configurator2.default && _configurator2.default.libraryName) libraryName = _configurator2.default.libraryName;else if ((0, _keys2.default)(_configurator2.default.injection.internal.normalized).length > 1) libraryName = '[name]';else {
    libraryName = _configurator2.default.name;
    if (_configurator2.default.exportFormat.self === 'var') libraryName = _clientnode2.default.stringConvertToValidVariableName(libraryName);
}
// / endregion
// / region plugins
const pluginInstances = [new _webpack2.default.NoEmitOnErrorsPlugin(), new _webpack2.default.optimize.OccurrenceOrderPlugin(true)];
if (_configurator2.default.debug) pluginInstances.push(new _webpack2.default.NamedModulesPlugin());
// // region define modules to ignore
for (const ignorePattern of _configurator2.default.injection.ignorePattern) pluginInstances.push(new _webpack2.default.IgnorePlugin(new RegExp(ignorePattern)));
// // endregion
// // region define modules to replace
for (const source in _configurator2.default.module.replacements.normal) if (_configurator2.default.module.replacements.normal.hasOwnProperty(source)) pluginInstances.push(new _webpack2.default.NormalModuleReplacementPlugin(new RegExp(source), _configurator2.default.module.replacements.normal[source]));
// // endregion
// // region generate html file
let htmlAvailable = false;
if (_configurator2.default.givenCommandLineArguments[2] !== 'build:dll') for (let htmlConfiguration of _configurator2.default.files.html) if (_clientnode2.default.isFileSync(htmlConfiguration.template.filePath)) {
    pluginInstances.push(new plugins.HTML(_clientnode2.default.extendObject({}, htmlConfiguration, {
        template: htmlConfiguration.template.request })));
    htmlAvailable = true;
}
// // endregion
// // region generate favicons
if (htmlAvailable && _configurator2.default.favicon && _clientnode2.default.isFileSync(_configurator2.default.favicon.logo)) pluginInstances.push(new plugins.Favicon(_configurator2.default.favicon));
// // endregion
// // region provide offline functionality
if (htmlAvailable && _configurator2.default.offline) {
    if (!['serve', 'test:browser'].includes(_configurator2.default.givenCommandLineArguments[2])) {
        if (_configurator2.default.inPlace.cascadingStyleSheet && (0, _keys2.default)(_configurator2.default.inPlace.cascadingStyleSheet).length) _configurator2.default.offline.excludes.push(_path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.path.target.asset.cascadingStyleSheet) + `*.css?${_configurator2.default.hashAlgorithm}=*`);
        if (_configurator2.default.inPlace.javaScript && (0, _keys2.default)(_configurator2.default.inPlace.javaScript).length) _configurator2.default.offline.excludes.push(_path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.path.target.asset.javaScript) + `*.js?${_configurator2.default.hashAlgorithm}=*`);
    }
    pluginInstances.push(new plugins.Offline(_configurator2.default.offline));
}
// // endregion
// // region opens browser automatically
if (_configurator2.default.development.openBrowser && htmlAvailable && ['serve', 'test:browser'].includes(_configurator2.default.givenCommandLineArguments[2])) pluginInstances.push(new plugins.OpenBrowser(_configurator2.default.development.openBrowser));
// // endregion
// // region provide build environment
if (_configurator2.default.build.definitions) pluginInstances.push(new _webpack2.default.DefinePlugin(_configurator2.default.build.definitions));
if (_configurator2.default.module.provide) pluginInstances.push(new _webpack2.default.ProvidePlugin(_configurator2.default.module.provide));
// // endregion
// // region modules/assets
// /// region perform javaScript minification/optimisation
if (_configurator2.default.module.optimizer.babili) pluginInstances.push(new _babiliWebpackPlugin2.default(_configurator2.default.module.optimizer.babili));
// /// endregion
// /// region apply module pattern
pluginInstances.push({ apply: compiler => {
        compiler.plugin('emit', (compilation, callback) => {
            for (const request in compilation.assets) if (compilation.assets.hasOwnProperty(request)) {
                const filePath = request.replace(/\?[^?]+$/, '');
                const type = _helper2.default.determineAssetType(filePath, _configurator2.default.build.types, _configurator2.default.path);
                if (type && _configurator2.default.assetPattern[type] && !new RegExp(_configurator2.default.assetPattern[type].excludeFilePathRegularExpression).test(filePath)) {
                    const source = compilation.assets[request].source();
                    if (typeof source === 'string') compilation.assets[request] = new _webpackSources.RawSource(_configurator2.default.assetPattern[type].pattern.replace(/\{1\}/g, source.replace(/\$/g, '$$$')));
                }
            }
            callback();
        });
    } });
// /// endregion
// /// region in-place configured assets in the main html file
if (htmlAvailable && !['serve', 'test:browser'].includes(_configurator2.default.givenCommandLineArguments[2])) pluginInstances.push({ apply: compiler => {
        const filePathsToRemove = [];
        compiler.plugin('compilation', compilation => compilation.plugin('html-webpack-plugin-after-html-processing', (() => {
            var _ref = (0, _asyncToGenerator3.default)(function* (htmlPluginData, callback) {
                if (_configurator2.default.inPlace.cascadingStyleSheet && (0, _keys2.default)(_configurator2.default.inPlace.cascadingStyleSheet).length || _configurator2.default.inPlace.javaScript && (0, _keys2.default)(_configurator2.default.inPlace.javaScript).length) try {
                    const result = yield _helper2.default.inPlaceCSSAndJavaScriptAssetReferences(htmlPluginData.html, _configurator2.default.inPlace.cascadingStyleSheet, _configurator2.default.inPlace.javaScript, _configurator2.default.path.target.base, _configurator2.default.files.compose.cascadingStyleSheet, _configurator2.default.files.compose.javaScript, compilation.assets);
                    htmlPluginData.html = result.content;
                    filePathsToRemove.concat(result.filePathsToRemove);
                } catch (error) {
                    return callback(error, htmlPluginData);
                }
                callback(null, htmlPluginData);
            });

            return function (_x, _x2) {
                return _ref.apply(this, arguments);
            };
        })()));
        compiler.plugin('after-emit', (() => {
            var _ref2 = (0, _asyncToGenerator3.default)(function* (compilation, callback) {
                let promises = [];
                for (const path of filePathsToRemove) if (yield _clientnode2.default.isFile(path)) promises.push(new _promise2.default(function (resolve) {
                    return fileSystem.unlink(path, function (error) {
                        if (error) console.error(error);
                        resolve();
                    });
                }));
                yield _promise2.default.all(promises);
                promises = [];
                for (const type of ['javaScript', 'cascadingStyleSheet']) promises.push(new _promise2.default(function (resolve, reject) {
                    return fileSystem.readdir(_configurator2.default.path.target.asset[type], _configurator2.default.encoding, function (error, files) {
                        if (error) {
                            reject(error);
                            return;
                        }
                        if (files.length === 0) fileSystem.rmdir(_configurator2.default.path.target.asset[type], function (error) {
                            return error ? reject(error) : resolve();
                        });else resolve();
                    });
                }));
                yield _promise2.default.all(promises);
                callback();
            });

            return function (_x3, _x4) {
                return _ref2.apply(this, arguments);
            };
        })());
    } });
// /// endregion
// /// region remove chunks if a corresponding dll package exists
if (_configurator2.default.givenCommandLineArguments[2] !== 'build:dll') for (const chunkName in _configurator2.default.injection.internal.normalized) if (_configurator2.default.injection.internal.normalized.hasOwnProperty(chunkName)) {
    const manifestFilePath = `${_configurator2.default.path.target.base}/${chunkName}.` + `dll-manifest.json`;
    if (_configurator2.default.dllManifestFilePaths.includes(manifestFilePath)) {
        delete _configurator2.default.injection.internal.normalized[chunkName];
        const filePath = _helper2.default.renderFilePathTemplate(_helper2.default.stripLoader(_configurator2.default.files.compose.javaScript), { '[name]': chunkName });
        pluginInstances.push(new plugins.AddAssetHTMLPlugin({
            filepath: filePath,
            hash: true,
            includeSourcemap: _clientnode2.default.isFileSync(`${filePath}.map`)
        }));
        pluginInstances.push(new _webpack2.default.DllReferencePlugin({
            context: _configurator2.default.path.context, manifest: require(manifestFilePath) }));
    }
}
// /// endregion
// /// region generate common chunks
if (_configurator2.default.givenCommandLineArguments[2] !== 'build:dll') for (const chunkName of _configurator2.default.injection.commonChunkIDs) if (_configurator2.default.injection.internal.normalized.hasOwnProperty(chunkName)) pluginInstances.push(new _webpack2.default.optimize.CommonsChunkPlugin({
    async: false,
    children: false,
    filename: _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.javaScript),
    minChunks: Infinity,
    name: chunkName,
    minSize: 0
}));
// /// endregion
// /// region mark empty javaScript modules as dummy
if (!_configurator2.default.needed.javaScript) _configurator2.default.files.compose.javaScript = _path2.default.resolve(_configurator2.default.path.target.asset.javaScript, '.__dummy__.compiled.js');
// /// endregion
// /// region extract cascading style sheets
if (_configurator2.default.files.compose.cascadingStyleSheet) pluginInstances.push(new plugins.ExtractText({
    allChunks: true, filename: _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.cascadingStyleSheet)
}));
// /// endregion
// /// region performs implicit external logic
if (_configurator2.default.injection.external.modules === '__implicit__')
    /*
        We only want to process modules from local context in library mode,
        since a concrete project using this library should combine all assets
        (and deduplicate them) for optimal bundling results. NOTE: Only native
        javaScript and json modules will be marked as external dependency.
    */
    _configurator2.default.injection.external.modules = (context, request, callback) => {
        request = request.replace(/^!+/, '');
        if (request.startsWith('/')) request = _path2.default.relative(_configurator2.default.path.context, request);
        for (const filePath of _configurator2.default.module.directoryNames.concat(_configurator2.default.loader.directoryNames)) if (request.startsWith(filePath)) {
            request = request.substring(filePath.length);
            if (request.startsWith('/')) request = request.substring(1);
            break;
        }
        let resolvedRequest = _helper2.default.determineExternalRequest(request, _configurator2.default.path.context, context, _configurator2.default.injection.internal.normalized, _configurator2.default.path.ignore.concat(_configurator2.default.module.directoryNames, _configurator2.default.loader.directoryNames).map(filePath => _path2.default.resolve(_configurator2.default.path.context, filePath)).filter(filePath => !_configurator2.default.path.context.startsWith(filePath)), _configurator2.default.module.aliases, _configurator2.default.module.replacements.normal, _configurator2.default.extensions, _configurator2.default.path.source.asset.base, _configurator2.default.path.ignore, _configurator2.default.module.directoryNames, _configurator2.default.package.main.fileNames, _configurator2.default.package.main.propertyNames, _configurator2.default.package.aliasPropertyNames, _configurator2.default.injection.external.implicit.pattern.include, _configurator2.default.injection.external.implicit.pattern.exclude, _configurator2.default.inPlace.externalLibrary.normal, _configurator2.default.inPlace.externalLibrary.dynamic, _configurator2.default.encoding);
        if (resolvedRequest) {
            if (['var', 'umd'].includes(_configurator2.default.exportFormat.external) && request in _configurator2.default.injection.external.aliases) resolvedRequest = _configurator2.default.injection.external.aliases[request];
            if (_configurator2.default.exportFormat.external === 'var') resolvedRequest = _clientnode2.default.stringConvertToValidVariableName(resolvedRequest, '0-9a-zA-Z_$\\.');
            return callback(null, resolvedRequest, _configurator2.default.exportFormat.external);
        }
        return callback();
    };
// /// endregion
// /// region build dll packages
if (_configurator2.default.givenCommandLineArguments[2] === 'build:dll') {
    let dllChunkIDExists = false;
    for (const chunkName in _configurator2.default.injection.internal.normalized) if (_configurator2.default.injection.internal.normalized.hasOwnProperty(chunkName)) if (_configurator2.default.injection.dllChunkIDs.includes(chunkName)) dllChunkIDExists = true;else delete _configurator2.default.injection.internal.normalized[chunkName];
    if (dllChunkIDExists) {
        libraryName = '[name]DLLPackage';
        pluginInstances.push(new _webpack2.default.DllPlugin({
            path: `${_configurator2.default.path.target.base}/[name].dll-manifest.json`,
            name: libraryName
        }));
    } else console.warn('No dll chunk id found.');
}
// /// endregion
// // endregion
// // region apply final dom/javaScript modifications/fixes
pluginInstances.push({ apply: compiler => compiler.plugin('compilation', compilation => {
        compilation.plugin('html-webpack-plugin-alter-asset-tags', (htmlPluginData, callback) => {
            for (const tags of [htmlPluginData.body, htmlPluginData.head]) {
                let index = 0;
                for (const tag of tags) {
                    if (/^\.__dummy__(\..*)?$/.test(_path2.default.basename(tag.attributes.src || tag.attributes.href || ''))) tags.splice(index, 1);
                    index += 1;
                }
            }
            const assets = JSON.parse(htmlPluginData.plugin.assetJson);
            let index = 0;
            for (const assetRequest of assets) {
                if (/^\.__dummy__(\..*)?$/.test(_path2.default.basename(assetRequest))) assets.splice(index, 1);
                index += 1;
            }
            htmlPluginData.plugin.assetJson = (0, _stringify2.default)(assets);
            callback(null, htmlPluginData);
        });
        compilation.plugin('html-webpack-plugin-after-html-processing', (htmlPluginData, callback) => {
            let window;
            try {
                window = new _jsdom.JSDOM(htmlPluginData.html.replace(/<%/g, '##+#+#+##').replace(/%>/g, '##-#-#-##')).window;
            } catch (error) {
                return callback(error, htmlPluginData);
            }
            const linkables = {
                script: 'src', link: 'href' };
            for (const tagName in linkables) if (linkables.hasOwnProperty(tagName)) for (const domNode of window.document.querySelectorAll(`${tagName}[${linkables[tagName]}*="?` + `${_configurator2.default.hashAlgorithm}="]`))
            /*
                NOTE: Removing symbols after a "&" in hash string
                is necessary to match the generated request strings
                in offline plugin.
            */
            domNode.setAttribute(linkables[tagName], domNode.getAttribute(linkables[tagName]).replace(new RegExp(`(\\?${_configurator2.default.hashAlgorithm}=` + '[^&]+).*$'), '$1'));
            htmlPluginData.html = htmlPluginData.html.replace(/^(\s*<!doctype [^>]+?>\s*)[\s\S]*$/i, '$1') + window.document.documentElement.outerHTML.replace(/##\+#\+#\+##/g, '<%').replace(/##-#-#-##/g, '%>');
            //  region post compilation
            for (const htmlFileSpecification of _configurator2.default.files.html) if (htmlFileSpecification.filename === htmlPluginData.plugin.options.filename) {
                for (const loaderConfiguration of htmlFileSpecification.template.use) if (loaderConfiguration.hasOwnProperty('options') && loaderConfiguration.options.hasOwnProperty('compileSteps') && typeof loaderConfiguration.options.compileSteps === 'number') htmlPluginData.html = _ejsLoader2.default.bind(_clientnode2.default.extendObject(true, {}, {
                    options: loaderConfiguration.options
                }, { options: {
                        compileSteps: htmlFileSpecification.template.postCompileSteps
                    } }))(htmlPluginData.html);
                break;
            }
            // endregion
            callback(null, htmlPluginData);
        });
    }) });
/*
    NOTE: The umd module export doesn't handle cases where the package name
    doesn't match exported library name. This post processing fixes this issue.
*/
if (_configurator2.default.exportFormat.external.startsWith('umd')) pluginInstances.push({ apply: compiler => compiler.plugin('emit', (compilation, callback) => {
        const bundleName = typeof libraryName === 'string' ? libraryName : libraryName[0];
        for (const assetRequest in compilation.assets) if (compilation.assets.hasOwnProperty(assetRequest) && assetRequest.replace(/([^?]+)\?.*$/, '$1').endsWith(_configurator2.default.build.types.javaScript.outputExtension)) {
            let source = compilation.assets[assetRequest].source();
            if (typeof source === 'string') {
                for (const replacement in _configurator2.default.injection.external.aliases) if (_configurator2.default.injection.external.aliases.hasOwnProperty(replacement)) source = source.replace(new RegExp('(require\\()"' + _clientnode2.default.stringEscapeRegularExpressions(_configurator2.default.injection.external.aliases[replacement]) + '"(\\))', 'g'), `$1'${replacement}'$2`).replace(new RegExp('(define\\("' + _clientnode2.default.stringEscapeRegularExpressions(bundleName) + '", \\[.*)"' + _clientnode2.default.stringEscapeRegularExpressions(_configurator2.default.injection.external.aliases[replacement]) + '"(.*\\], factory\\);)'), `$1'${replacement}'$2`);
                source = source.replace(new RegExp('(root\\[)"' + _clientnode2.default.stringEscapeRegularExpressions(bundleName) + '"(\\] = )'), `$1'` + _clientnode2.default.stringConvertToValidVariableName(bundleName) + `'$2`);
                compilation.assets[assetRequest] = new _webpackSources.RawSource(source);
            }
        }
        callback();
    }) });
// // endregion
// // region add automatic image compression
// NOTE: This plugin should be loaded at last to ensure that all emitted images
// ran through.
pluginInstances.push(new plugins.Imagemin(_configurator2.default.module.optimizer.image.content));
// // endregion
// // region context replacements
for (const contextReplacement of _configurator2.default.module.replacements.context) pluginInstances.push(new _webpack2.default.ContextReplacementPlugin(...contextReplacement.map(value => new Function('configuration', '__dirname', '__filename', `return ${value}`
// IgnoreTypeCheck
)(_configurator2.default, __dirname, __filename))));
// // endregion
// / endregion
// / region loader helper
const rejectFilePathInDependencies = filePath => {
    filePath = _helper2.default.stripLoader(filePath);
    return _helper2.default.isFilePathInLocation(filePath, _configurator2.default.path.ignore.concat(_configurator2.default.module.directoryNames, _configurator2.default.loader.directoryNames).map(filePath => _path2.default.resolve(_configurator2.default.path.context, filePath)).filter(filePath => !_configurator2.default.path.context.startsWith(filePath)));
};
const loader = {};
const evaluate = (code, filePath) => new Function('configuration', 'filePath', 'loader', 'rejectFilePathInDependencies', `return ${code}`
// IgnoreTypeCheck
)(_configurator2.default, filePath, loader, rejectFilePathInDependencies);
_clientnode2.default.extendObject(loader, {
    // Convert to compatible native web types.
    // region generic template
    ejs: {
        exclude: filePath => _helper2.default.normalizePaths(_configurator2.default.files.html.concat(_configurator2.default.files.defaultHTML).map(htmlConfiguration => htmlConfiguration.template.filePath)).includes(filePath) || (_configurator2.default.module.preprocessor.ejs.exclude === null ? false : evaluate(_configurator2.default.module.preprocessor.ejs.exclude, filePath)),
        include: _helper2.default.normalizePaths([_configurator2.default.path.source.base].concat(_configurator2.default.module.locations.directoryPaths)),
        test: /^(?!.+\.html\.ejs$).+\.ejs$/i,
        use: [{ loader: 'file?name=[path][name]' + (Boolean(_configurator2.default.module.preprocessor.ejs.options.compileSteps % 2) ? '.js' : '') + `?${_configurator2.default.hashAlgorithm}=[hash]` }, { loader: 'extract' }, {
            loader: _configurator2.default.module.preprocessor.ejs.loader,
            options: _configurator2.default.module.preprocessor.ejs.options
        }].concat(_configurator2.default.module.preprocessor.ejs.additional.map(evaluate))
    },
    // endregion
    // region script
    script: {
        exclude: filePath => _configurator2.default.module.preprocessor.javaScript.exclude === null ? rejectFilePathInDependencies(filePath) : evaluate(_configurator2.default.module.preprocessor.javaScript.exclude, filePath),
        include: _helper2.default.normalizePaths([_configurator2.default.path.source.asset.javaScript].concat(_configurator2.default.module.locations.directoryPaths)),
        test: /\.js(?:\?.*)?$/i,
        use: [{
            loader: _configurator2.default.module.preprocessor.javaScript.loader,
            options: _configurator2.default.module.preprocessor.javaScript.options
        }].concat(_configurator2.default.module.preprocessor.javaScript.additional.map(evaluate))
    },
    // endregion
    // region html template
    html: {
        // NOTE: This is only for the main entry template.
        main: {
            test: new RegExp('^' + _clientnode2.default.stringEscapeRegularExpressions(_configurator2.default.files.defaultHTML.template.filePath) + '(?:\\?.*)?$'),
            use: _configurator2.default.files.defaultHTML.template.use
        },
        ejs: {
            exclude: filePath => _helper2.default.normalizePaths(_configurator2.default.files.html.concat(_configurator2.default.files.defaultHTML).map(htmlConfiguration => htmlConfiguration.template.filePath)).includes(filePath) || (_configurator2.default.module.preprocessor.html.exclude === null ? false : evaluate(_configurator2.default.module.preprocessor.html.exclude, filePath)),
            include: _configurator2.default.path.source.asset.template,
            test: /\.html\.ejs(?:\?.*)?$/i,
            use: [{ loader: 'file?name=' + _path2.default.join(_path2.default.relative(_configurator2.default.path.target.asset.base, _configurator2.default.path.target.asset.template), '[name]' + (Boolean(_configurator2.default.module.preprocessor.html.options.compileSteps % 2) ? '.js' : '') + `?${_configurator2.default.hashAlgorithm}=[hash]`) }].concat(Boolean(_configurator2.default.module.preprocessor.html.options.compileSteps % 2) ? [] : [{ loader: 'extract' }, {
                loader: _configurator2.default.module.html.loader,
                options: _configurator2.default.module.html.options
            }], {
                loader: _configurator2.default.module.preprocessor.html.loader,
                options: _configurator2.default.module.preprocessor.html.options
            }).concat(_configurator2.default.module.preprocessor.html.additional.map(evaluate))
        },
        html: {
            exclude: filePath => _helper2.default.normalizePaths(_configurator2.default.files.html.concat(_configurator2.default.files.defaultHTML).map(htmlConfiguration => htmlConfiguration.template.filePath)).includes(filePath) || (_configurator2.default.module.html.exclude === null ? true : evaluate(_configurator2.default.module.html.exclude, filePath)),
            include: _configurator2.default.path.source.asset.template,
            test: /\.html(?:\?.*)?$/i,
            use: [{ loader: 'file?name=' + _path2.default.join(_path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.path.target.asset.template), `[name].[ext]?${_configurator2.default.hashAlgorithm}=[hash]`) }, { loader: 'extract' }, {
                loader: _configurator2.default.module.html.loader,
                options: _configurator2.default.module.html.options
            }].concat(_configurator2.default.module.html.additional.map(evaluate))
        }
    },
    // endregion
    // Load dependencies.
    // region style
    style: {
        exclude: filePath => _configurator2.default.module.cascadingStyleSheet.exclude === null ? rejectFilePathInDependencies(filePath) : evaluate(_configurator2.default.module.cascadingStyleSheet.exclude, filePath),
        include: _helper2.default.normalizePaths([_configurator2.default.path.source.asset.cascadingStyleSheet].concat(_configurator2.default.module.locations.directoryPaths)),
        test: /\.s?css(?:\?.*)?$/i,
        use: [{
            loader: _configurator2.default.module.style.loader,
            options: _configurator2.default.module.style.options
        }, {
            loader: _configurator2.default.module.cascadingStyleSheet.loader,
            options: _configurator2.default.module.cascadingStyleSheet.options
        }, {
            loader: _configurator2.default.module.preprocessor.cascadingStyleSheet.loader,
            options: _clientnode2.default.extendObject(true, {
                ident: 'postcss',
                plugins: () => [(0, _postcssImport2.default)({
                    addDependencyTo: _webpack2.default,
                    root: _configurator2.default.path.context
                }), (0, _postcssCssnext2.default)({ browsers: '> 0%' }),
                /*
                    NOTE: Checking path doesn't work if fonts are
                    referenced in libraries provided in another
                    location than the project itself like the
                    "node_modules" folder.
                */
                (0, _postcssFontpath2.default)({ checkPath: false }), (0, _postcssUrl2.default)({ url: 'rebase' }), (0, _postcssSprites2.default)({
                    filterBy: () => new _promise2.default((resolve, reject) => (_configurator2.default.files.compose.image ? resolve : reject)()),
                    hooks: { onSaveSpritesheet: image => _path2.default.join(image.spritePath, _path2.default.relative(_configurator2.default.path.target.asset.image, _configurator2.default.files.compose.image))
                    },
                    stylesheetPath: _configurator2.default.path.source.asset.cascadingStyleSheet,
                    spritePath: _configurator2.default.path.source.asset.image
                })]
            }, _configurator2.default.module.preprocessor.cascadingStyleSheet.options)
        }].concat(_configurator2.default.module.preprocessor.cascadingStyleSheet.additional.map(evaluate))
    },
    // endregion
    // Optimize loaded assets.
    // region font
    font: {
        eot: {
            exclude: filePath => _configurator2.default.module.optimizer.font.eot.exclude === null ? false : evaluate(_configurator2.default.module.optimizer.font.eot.exclude, filePath),
            include: _configurator2.default.path.base,
            test: /\.eot(?:\?.*)?$/i,
            use: [{
                loader: _configurator2.default.module.optimizer.font.eot.loader,
                options: _configurator2.default.module.optimizer.font.eot.options
            }].concat(_configurator2.default.module.optimizer.font.eot.additional.map(evaluate))
        },
        svg: {
            exclude: filePath => _configurator2.default.module.optimizer.font.svg.exclude === null ? false : evaluate(_configurator2.default.module.optimizer.font.svg.exclude, filePath),
            include: _configurator2.default.path.base,
            test: /\.svg(?:\?.*)?$/i,
            use: [{
                loader: _configurator2.default.module.optimizer.font.svg.loader,
                options: _configurator2.default.module.optimizer.font.svg.options
            }].concat(_configurator2.default.module.optimizer.font.svg.additional.map(evaluate))
        },
        ttf: {
            exclude: filePath => _configurator2.default.module.optimizer.font.ttf.exclude === null ? false : evaluate(_configurator2.default.module.optimizer.font.ttf.exclude, filePath),
            include: _configurator2.default.path.base,
            test: /\.ttf(?:\?.*)?$/i,
            use: [{
                loader: _configurator2.default.module.optimizer.font.ttf.loader,
                options: _configurator2.default.module.optimizer.font.ttf.options
            }].concat(_configurator2.default.module.optimizer.font.ttf.additional.map(evaluate))
        },
        woff: {
            exclude: filePath => _configurator2.default.module.optimizer.font.woff.exclude === null ? false : evaluate(_configurator2.default.module.optimizer.font.woff.exclude, filePath),
            include: _configurator2.default.path.base,
            test: /\.woff2?(?:\?.*)?$/i,
            use: [{
                loader: _configurator2.default.module.optimizer.font.woff.loader,
                options: _configurator2.default.module.optimizer.font.woff.options
            }].concat(_configurator2.default.module.optimizer.font.woff.additional.map(evaluate))
        }
    },
    // endregion
    // region image
    image: {
        exclude: filePath => _configurator2.default.module.optimizer.image.exclude === null ? rejectFilePathInDependencies(filePath) : evaluate(_configurator2.default.module.optimizer.image.exclude, filePath),
        include: _configurator2.default.path.source.asset.image,
        test: /\.(?:png|jpg|ico|gif)(?:\?.*)?$/i,
        use: [{
            loader: _configurator2.default.module.optimizer.image.loader,
            options: _configurator2.default.module.optimizer.image.file
        }].concat(_configurator2.default.module.optimizer.image.additional.map(evaluate))
    },
    // endregion
    // region data
    data: {
        exclude: filePath => _configurator2.default.extensions.file.internal.includes(_path2.default.extname(_helper2.default.stripLoader(filePath))) || (_configurator2.default.module.optimizer.data.exclude === null ? rejectFilePathInDependencies(filePath) : evaluate(_configurator2.default.module.optimizer.data.exclude, filePath)),
        include: _configurator2.default.path.source.asset.data,
        test: /.+/,
        use: [{
            loader: _configurator2.default.module.optimizer.data.loader,
            options: _configurator2.default.module.optimizer.data.options
        }].concat(_configurator2.default.module.optimizer.data.additional.map(evaluate))
        // endregion
    } });
if (_configurator2.default.files.compose.cascadingStyleSheet) {
    loader.style.use.shift();
    loader.style.use = plugins.ExtractText.extract({ use: loader.style.use });
}
// / endregion
// endregion
// region configuration
const webpackConfiguration = {
    bail: true,
    cache: _configurator2.default.cache.main,
    context: _configurator2.default.path.context,
    devtool: _configurator2.default.development.tool,
    devServer: _configurator2.default.development.server,
    // region input
    entry: _configurator2.default.injection.internal.normalized,
    externals: _configurator2.default.injection.external.modules,
    resolve: {
        alias: _configurator2.default.module.aliases,
        aliasFields: _configurator2.default.package.aliasPropertyNames,
        extensions: _configurator2.default.extensions.file.internal,
        mainFields: _configurator2.default.package.main.propertyNames,
        mainFiles: _configurator2.default.package.main.fileNames,
        moduleExtensions: _configurator2.default.extensions.module,
        modules: _helper2.default.normalizePaths(_configurator2.default.module.directoryNames),
        unsafeCache: _configurator2.default.cache.unsafe
    },
    resolveLoader: {
        alias: _configurator2.default.loader.aliases,
        aliasFields: _configurator2.default.package.aliasPropertyNames,
        extensions: _configurator2.default.loader.extensions.file,
        mainFields: _configurator2.default.package.main.propertyNames,
        mainFiles: _configurator2.default.package.main.fileNames,
        moduleExtensions: _configurator2.default.loader.extensions.module,
        modules: _configurator2.default.loader.directoryNames
    },
    // endregion
    // region output
    output: {
        filename: _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.javaScript),
        hashFunction: _configurator2.default.hashAlgorithm,
        library: libraryName,
        libraryTarget: _configurator2.default.givenCommandLineArguments[2] === 'build:dll' ? 'var' : _configurator2.default.exportFormat.self,
        path: _configurator2.default.path.target.base,
        publicPath: _configurator2.default.path.target.public,
        pathinfo: _configurator2.default.debug,
        umdNamedDefine: true
    },
    performance: _configurator2.default.performanceHints,
    target: _configurator2.default.targetTechnology,
    // endregion
    module: {
        rules: _configurator2.default.module.additional.map(loaderConfiguration => {
            return {
                exclude: filePath => evaluate(loaderConfiguration.exclude || 'false', filePath),
                include: loaderConfiguration.include && evaluate(loaderConfiguration.include, _configurator2.default.path.context) || _configurator2.default.path.source.base,
                test: new RegExp(evaluate(loaderConfiguration.test, _configurator2.default.path.context)),
                use: evaluate(loaderConfiguration.use)
            };
        }).concat([loader.ejs, loader.script, loader.html.main, loader.html.ejs, loader.html.html, loader.style, loader.font.eot, loader.font.svg, loader.font.ttf, loader.font.woff, loader.image, loader.data])
    },
    plugins: pluginInstances
};
if (!Array.isArray(_configurator2.default.module.skipParseRegularExpressions) || _configurator2.default.module.skipParseRegularExpressions.length) webpackConfiguration.module.noParse = _configurator2.default.module.skipParseRegularExpressions;
if (_configurator2.default.showConfiguration) {
    console.info('Using internal configuration:', _util2.default.inspect(_configurator2.default, {
        depth: null }));
    console.info('-----------------------------------------------------------');
    console.info('Using webpack configuration:', _util2.default.inspect(webpackConfiguration, { depth: null }));
}
// endregion
exports.default = webpackConfiguration;
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion
